akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -491,3 +495,61 @@ private[akka] final case class MapAsyncUnordered[In, Out](parallelism: Int, f: I
     if (todo > 0) ctx.absorbTermination()
     else ctx.finish()
 }
+
+/**
+ * INTERNAL API
+ */
+private[akka] final case class Log[T](name: String, extract: T ⇒ Any, logAdapter: Option[LoggingAdapter]) extends PushStage[T, T] {
+  import Log._
+  private var log: LoggingAdapter = _
+  private var logLevels: OperationAttributes.LogLevels = _
+
+  override def onPush(elem: T, ctx: Context[T]): SyncDirective = {
+    initLogger(ctx) // TODO if we have preStart this could be called in it, instead of in all callbacks
+
+    log.log(logLevels.onElement, "[{}] Element: {}", name, extract(elem))
+    ctx.push(elem)
+  }
+
+  override def onUpstreamFailure(cause: Throwable, ctx: Context[T]): TerminationDirective = {
+    initLogger(ctx) // TODO if we have preStart this could be called in it, instead of in all callbacks
+
+    logLevels.onFailure match {
+      case Logging.ErrorLevel ⇒ log.error(cause, "[{}] Upstream failed.", name)
+      case level              ⇒ log.log(level, "[{}] Upstream failed, cause: {}: {}", name, Logging.simpleName(cause.getClass), cause.getMessage)
+    }
+    super.onUpstreamFailure(cause, ctx)
+  }
+
+  override def onUpstreamFinish(ctx: Context[T]): TerminationDirective = {
+    initLogger(ctx) // TODO if we have preStart this could be called in it, instead of in all callbacks
+
+    log.log(logLevels.onFinish, "[{}] Upstream finished.", name)
+    super.onUpstreamFinish(ctx)
+  }
+
+  override def onDownstreamFinish(ctx: Context[T]): TerminationDirective = {
```

Comment:
I'm also adding `OffLevel` support here - seen it used in http, where they only want to log failures for example.